plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java

```diff
@@ -499,6 +510,64 @@ public MongoCursor<Document> execute(MongoTableHandle tableHandle, List<MongoCol
         return iterable.iterator();
     }
 
+    /**
+     * Creates a set of sufficient columns for the input projected columns. For example,
+     * if input {@param columns} include columns "a.b" and "a.b.c", then they will be projected from a single column "a.b".
+     */
+    public static List<String> projectSufficientColumns(List<MongoColumnHandle> columnHandles)
+    {
+        List<MongoColumnHandle> modifiedColumnHandles = columnHandles.stream()
+                .map(column -> new MongoColumnHandle(
+                        column.getBaseName(),
+                        getModifiedDereferenceChain(column),
+                        column.getType(),
+                        column.isHidden(),
+                        column.isDbRefField(),
+                        column.getComment()))
+                .collect(toImmutableList());
+
+        List<MongoColumnHandle> sortedColumnHandles = COLUMN_HANDLE_ORDERING.sortedCopy(modifiedColumnHandles);
+        List<MongoColumnHandle> sufficientColumns = new ArrayList<>();
+        for (MongoColumnHandle column : sortedColumnHandles) {
+            if (!parentColumnExists(sufficientColumns, column)) {
+                sufficientColumns.add(column);
+            }
+        }
+        return sufficientColumns.stream()
+                .map(MongoColumnHandle::getQualifiedName)
+                .collect(toImmutableList());
+    }
+
+    private static List<String> getModifiedDereferenceChain(MongoColumnHandle columnHandle)
```

Comment:
Thanks a lot, @ebyhr for the suggestion. It really simplifies the code.

Instead of copying the method, I tried to pass `fieldDereferencePredicate` to `extractSupportedProjectedColumns`.